### PR TITLE
Republish the fixed apache extensions as google-http-client-apache-v2

### PR DESCRIPTION
**Pull request by chingor13**
1.29.0 reverted the changes that split the `google-http-client-apache` artifact out of `google-http-client`. Rather than "breaking" the original artifact, we will provide a new artifact `google-http-client-apache-v2` which provides a new package `com.google.api.client.http.apache.v2`.

This also removes the `google-http-client-apache` and `google-http-client-apache-legacy` modules. The legacy code still exists in the base `google-http-client` artifact and is now deprecated.

The new artifact will be versioned the same as the rest of the packages (so will start at 1.30.0 when first released).

### CODE REVIEW

File: google-http-client-apache-v2/src/main/java/com/google/api/client/http/apache/v2/ApacheHttpTransport.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    } else if (method.equals(HttpMethods.OPTIONS)) {
+      requestBase = new HttpOptions(url);
+    } else {
+      requestBase = new HttpExtensionMethod(method, url);
+    }
+    return new ApacheHttpRequest(httpClient, requestBase);
+  }
+
+  /**
+   * Shuts down the connection manager and releases allocated resources. This includes closing all
```
Comment (elharo) [inline google-http-client-apache-v2/src/main/java/com/google/api/client/http/apache/v2/ApacheHttpTransport.java]: includes closing --> closes

File: google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2011 Google Inc.
```
Comment (elharo) [inline google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java]: Copyright 2019 Google LLC.

File: google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  private void checkHttpClient(HttpClient client) {
+    assertNotNull(client);
+    // TODO(chingor): Is it possible to test this effectively? The newer HttpClient implementations
+    // are read-only and we're testing that we built the client with the right configuration
+  }
+
+  @Test
+  public void testRequestsWithContent() throws Exception {
```
Comment (elharo) [inline google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java]: Personally I prefer to be explicit about which exceptions a test might throw.

File: google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    HttpClient mockClient = mock(HttpClient.class);
+    HttpResponse mockResponse = mock(HttpResponse.class);
+    when(mockClient.execute(any(HttpUriRequest.class))).thenReturn(mockResponse);
+
+    ApacheHttpTransport transport = new ApacheHttpTransport(mockClient);
+
+    // Test GET.
```
Comment (elharo) [inline google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java]: These should probably be seven separate methods with a common setup here.

File: google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+      fail("expected " + IllegalArgumentException.class);
+    } catch (IllegalArgumentException e) {
+      // expected
+      assertEquals(e.getMessage(),
+          "Apache HTTP client does not support " + method + " requests with content.");
+    }
+  }
+
+  private void execute(ApacheHttpRequest request) throws Exception {
+    byte[] bytes = StringUtils.getBytesUtf8("abc");
```
Comment (elharo) [inline google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java]: Outside the scope of this PR, but StringUtils was a bad idea. This line is trivial to implement without an extra dependency.

File: google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    request.setContentLength(bytes.length);
+    request.execute();
+  }
+
+  @Test
+  public void testRequestShouldNotFollowRedirects() throws IOException {
+    final AtomicInteger requestsAttempted = new AtomicInteger(0);
+    HttpRequestExecutor requestExecutor = new HttpRequestExecutor() {
+      @Override
+      public HttpResponse execute(HttpRequest request, HttpClientConnection conn,
```
Comment (elharo) [inline google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java]: conn --> connection

File: google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+          HttpContext context) throws IOException, HttpException {
+        HttpResponse resp = new BasicHttpResponse(HttpVersion.HTTP_1_1, 302, null);
```
Comment (elharo) [inline google-http-client-apache-v2/src/test/java/com/google/api/client/http/apache/v2/ApacheHttpTransportTest.java]: resp --> response

File: google-http-client-apache-v2/pom.xml [hunk 8]
```diff
@@ -7,9 +7,9 @@
     <version>1.29.2-SNAPSHOT</version><!-- {x-version-update:google-http-client-parent:current} -->
     <relativePath>../pom.xml</relativePath>
   </parent>
-  <artifactId>google-http-client-apache</artifactId>
-  <version>2.1.2-SNAPSHOT</version><!-- {x-version-update:google-http-client-apache:current} -->
-  <name>Apache HTTP transport for the Google HTTP Client Library for Java.</name>
+  <artifactId>google-http-client-apache-v2</artifactId>
+  <version>1.29.2-SNAPSHOT</version><!-- {x-version-update:google-http-client-apache-v2:current} -->
```
[the remark this answers is not in the document]
Comment (chingor13) [reply google-http-client-apache-v2/pom.xml]: Our release tools will correctly version bump all of the artifacts. This should be included in the 1.30.0 release.

File: google-http-client-apache-v2/src/main/java/com/google/api/client/http/apache/v2/ApacheHttpRequest.java [hunk 9]
```diff
@@ -12,28 +12,31 @@
...
   private final HttpClient httpClient;
 
   private final HttpRequestBase request;
 
+  private RequestConfig.Builder requestConfig;
+
   ApacheHttpRequest(HttpClient httpClient, HttpRequestBase request) {
     this.httpClient = httpClient;
     this.request = request;
+    this.requestConfig = RequestConfig.custom().setRedirectsEnabled(false);
```
[the remark this answers is not in the document]
Comment (chingor13) [reply google-http-client-apache-v2/src/main/java/com/google/api/client/http/apache/v2/ApacheHttpRequest.java]: This is more of an implementation detail. google-http-client configures redirects so we turn off the underlying automatic redirects in the Apache client.

**Pull request merged**